### PR TITLE
Cb/simulation results

### PR DESCRIPTION
**Pull request by claytonpbarrows**

### CODE REVIEW

File: src/operation/problem_results.jl [hunk 1]
```diff
@@ -411,32 +454,135 @@ function read_realized_aux_variables(res::ProblemResults)
...
+        variables;
+        initial_time = initial_time,
+        count = count,
     )
+    return Dict(encode_key_as_string(k) => v for (k, v) in result_values)
 end
 
-function read_realized_variables(res::ProblemResults)
-    return _read_realized_results(res.variable_values, nothing)
+function read_realized_variables_with_keys(
```
Comment (daniel-thom) [inline src/operation/problem_results.jl:490]: Is this function private to this file? If so, `_read_realized_variables_with_keys`

File: src/operation/problem_results.jl [hunk 2]
```diff
@@ -411,32 +454,135 @@ function read_realized_aux_variables(res::ProblemResults)
...
-    return _read_realized_results(
-        res.variable_values,
-        [VariableKey(x...) for x in variable_keys],
+"""
+Return the values for the requested variable keys for a problem.
+Accepts a vector of keys for the return of the values. If the time stamps and keys are
+loaded using the [load_results!](@ref) function it will read from memory.
+
+# Arguments
+- `variables::Vector{<:OptimizationContainerKey}` : keys of desired results
```
Comment (daniel-thom) [inline src/operation/problem_results.jl:463]: Docstring needs updating - don't refer to keys.

File: src/simulation/simulation_problem_results.jl [hunk 3]
```diff
@@ -152,6 +167,10 @@ Set the system in the results instance.
 
 Throws InvalidValue if the system UUID is incorrect.
 """
+function set_system!(results::SimulationProblemResults, system::AbstractString)
```
Comment (daniel-thom) [inline src/simulation/simulation_problem_results.jl:170]: Docstring should say that this is a path to the JSON file.

**Review by daniel-thom — COMMENTED** [review]
Looks good to me. My only real comment is that the docstrings for read_realized_* need to remove "keys" and give examples. I'd be willing to help to do this.

File: src/operation/problem_results.jl (hunk 1 above)
Comment (claytonpbarrows) [reply src/operation/problem_results.jl:490]: It's used in PowerGraphics. It definately doesn't need to be exported, but should I still prefix with "_"?

File: src/operation/problem_results.jl (hunk 2 above)
Comment (jd-lara) [reply src/operation/problem_results.jl]: Fixed via ed3223440a445d69e20ccb4f915c70c15bbeb82f

File: src/simulation/simulation_problem_results.jl (hunk 3 above)
Comment (jd-lara) [reply src/simulation/simulation_problem_results.jl:182]: fixed via 21a39e66bd2921ac995a2ac048537fee285b3f41

**Comment by jd-lara** [comment]
@claytonpbarrows the PR is passing tests now. The biggest missing item is that we don't have the interfaces to read `expressions` and `auxiliary_variables`. I think that if @daniel-thom approves the PR we should merge and address the missing results gathering in a forthcoming PR

File: src/core/constraints.jl [hunk 4]
```diff
@@ -107,3 +107,8 @@ struct ComponentActivePowerVariableLimitsConstraint <: PowerVariableLimitsConstr
 struct ComponentReactivePowerVariableLimitsConstraint <: PowerVariableLimitsConstraint end
 
 should_write_resulting_value(::Type{<:ConstraintType}) = true
+
+convert_result_to_natural_units(::Type{<:ConstraintType}) = false
+convert_result_to_natural_units(::Type{CopperPlateBalanceConstraint}) = true
```
Comment (jd-lara) [inline src/core/constraints.jl:112]: Why do you want conversion for these if the cost function is already in the MW-monetary unit. None of these should be converted when querying the duals.

File: src/core/parameters.jl [hunk 5]
```diff
@@ -185,3 +185,19 @@ abstract type AuxVariableValueParameter <: RightHandSideParameter end
...
 should_write_resulting_value(::Type{<:RightHandSideParameter}) = true
+
+convert_result_to_natural_units(::Type{<:ParameterType}) = false
+
+convert_result_to_natural_units(::Type{ActivePowerTimeSeriesParameter}) = true
+convert_result_to_natural_units(::Type{ReactivePowerTimeSeriesParameter}) = true
+convert_result_to_natural_units(::Type{RequirementTimeSeriesParameter}) = true
+convert_result_to_natural_units(::Type{EnergyTargetTimeSeriesParameter}) = true
+convert_result_to_natural_units(::Type{EnergyBudgetTimeSeriesParameter}) = true
+#convert_result_to_natural_units(::Type{InflowTimeSeriesParameter}) = true # TODO: is this pu?
```
Comment (jd-lara) [inline src/core/parameters.jl:196]: these ones will depend on the input data.

File: src/core/parameters.jl [hunk 6]
```diff
@@ -185,3 +185,19 @@ abstract type AuxVariableValueParameter <: RightHandSideParameter end
...
+#convert_result_to_natural_units(::Type{OutflowTimeSeriesParameter}) = true # TODO: is this pu?
+convert_result_to_natural_units(::Type{UpperBoundValueParameter}) = true
+convert_result_to_natural_units(::Type{LowerBoundValueParameter}) = true
+convert_result_to_natural_units(::Type{IntegralLimitParameter}) = true
+convert_result_to_natural_units(::Type{FixValueParameter}) = true
```
Comment (jd-lara) [inline src/core/parameters.jl:201]: This one is used to fix values of many kinds including on-off. Delete so its set to false.

File: src/core/parameters.jl [hunk 7]
```diff
@@ -185,3 +185,19 @@ abstract type AuxVariableValueParameter <: RightHandSideParameter end
...
+convert_result_to_natural_units(::Type{EnergyTargetParameter}) = true
+#convert_result_to_natural_units(::Type{CostFunctionParameter}) = true # TODO: is this pu?
```
Comment (jd-lara) [inline src/core/parameters.jl:203]: No, this is in MW-monetary values.

File: src/core/variables.jl [hunk 8]
```diff
@@ -135,4 +135,38 @@ const START_VARIABLES = (HotStartVariable, WarmStartVariable, ColdStartVariable)
...
+convert_result_to_natural_units(::Type{ActivePowerVariable}) = true
+convert_result_to_natural_units(::Type{PowerAboveMinimumVariable}) = true
+convert_result_to_natural_units(::Type{ActivePowerInVariable}) = true
+convert_result_to_natural_units(::Type{ActivePowerOutVariable}) = true
+convert_result_to_natural_units(::Type{EnergyVariable}) = true
+convert_result_to_natural_units(::Type{EnergyVariableUp}) = true
+convert_result_to_natural_units(::Type{EnergyVariableDown}) = true
+convert_result_to_natural_units(::Type{EnergyShortageVariable}) = true
+convert_result_to_natural_units(::Type{EnergySurplusVariable}) = true
+#convert_result_to_natural_units(::Type{LiftVariable }) = true # TODO: is this pu?
```
Comment (jd-lara) [inline src/core/variables.jl:149]: not in P.u. delete.

File: src/core/variables.jl [hunk 9]
```diff
@@ -135,4 +135,38 @@ const START_VARIABLES = (HotStartVariable, WarmStartVariable, ColdStartVariable)
...
+convert_result_to_natural_units(::Type{ReactivePowerVariable}) = true
+convert_result_to_natural_units(::Type{ActivePowerReserveVariable}) = true
+convert_result_to_natural_units(::Type{ServiceRequirementVariable}) = true
+#convert_result_to_natural_units(::Type{WaterSpillageVariable }) = true # TODO: is this pu?
+#convert_result_to_natural_units(::Type{AreaMismatchVariable }) = true # TODO: is this pu?
```
Comment (jd-lara) [inline src/core/variables.jl:154]: yes, this is in power

File: src/core/variables.jl [hunk 10]
```diff
@@ -135,4 +135,38 @@ const START_VARIABLES = (HotStartVariable, WarmStartVariable, ColdStartVariable)
...
+convert_result_to_natural_units(::Type{DeltaActivePowerUpVariable}) = true
+convert_result_to_natural_units(::Type{DeltaActivePowerDownVariable}) = true
+convert_result_to_natural_units(::Type{AdditionalDeltaActivePowerUpVariable}) = true
+convert_result_to_natural_units(::Type{AdditionalDeltaActivePowerDownVariable}) = true
+#convert_result_to_natural_units(::Type{SmoothACE }) = true # TODO: is this pu?
```
Comment (jd-lara) [inline src/core/variables.jl:159]: Yes, ACE is in Power.

File: src/operation/problem_results.jl [hunk 11]
```diff
@@ -31,13 +31,14 @@ list_expression_names(res::ProblemResults) =
 get_timestamps(res::ProblemResults) = res.timestamps
 get_model_base_power(res::ProblemResults) = res.base_power
 get_dual_values(res::ProblemResults) = res.dual_values
-get_expressionl_values(res::ProblemResults) = res.expression_values
+get_expression_values(res::ProblemResults) = res.expression_values
```
Comment (jd-lara) [inline src/operation/problem_results.jl:34]: missing `get_aux_variables_values`

File: src/core/auxiliary_variables.jl [hunk 12]
```diff
@@ -23,3 +23,5 @@ struct TimeDurationOff <: AuxVariableType end
 struct PowerOutput <: AuxVariableType end
```
Comment (jd-lara) [inline src/core/auxiliary_variables.jl:23]: needs conversion

File: src/operation/problem_results.jl [hunk 13]
```diff
@@ -377,28 +398,36 @@ _get_keys(::Type{VariableKey}, res, ::Nothing) = collect(keys(res.variable_value
...
     container_keys = container_keys === nothing ? existing_keys : container_keys
     _validate_keys(existing_keys, container_keys)
-    return Dict(
-        encode_key_as_string(k) => v for (k, v) in result_values if k in container_keys
+    return Dict{OptimizationContainerKey, DataFrames.DataFrame}(
+        k => v[time_ids, :] for (k, v) in result_values if k in container_keys
     )
 end
 
-# TODO: This is not used. Can it be deleted?
```
Comment (jd-lara) [inline src/operation/problem_results.jl:389]: @daniel-thom yeah, this isn't used anymore

File: src/operation/problem_results.jl (hunk 1 above)
Comment (daniel-thom) [reply src/operation/problem_results.jl:477]: Then I think it is good as-is.

**Review by daniel-thom — APPROVED** [review]

**Pull request merged**